File: KDesktopVirt/bindings/nodejs/kvirtualstage.js
/**
 * KVirtualStage Node.js Bindings
 * 
 * Playwright-equivalent desktop automation platform for AI agents.
 * Provides high-level JavaScript/TypeScript API for:
 * - Desktop automation and control  
 * - Session management
 * - Recording and playbook
 * - Natural human-like interactions
 * - Cross-platform virtualization
 * 
 * @example
 * ```javascript
 * const kvs = require('kvirtualstage');
 * 
 * // Create automation instance
 * const automation = new kvs.KVirtualStage();
 * 
 * // Create a new desktop session
 * const session = await automation.createSession({
 *   userId: 'demo_user',
 *   sessionName: 'my_session', 
 *   desktopType: 'ubuntu'
 * });
 * 
 * // Perform natural automation
 * await session.moveCursor(400, 300);
 * await session.click();
 * await session.typeText('Hello from KVirtualStage!');
 * 
 * // Start recording
 * const recording = await session.startRecording('demo.mp4');
 * 
 * // Execute workflow
 * const workflow = new kvs.Workflow('Calculator Demo');
 * workflow.addStep('moveCursor', { x: 100, y: 100 });
 * workflow.addStep('click');
 * workflow.addStep('type', { text: '2 + 2 =' });
 * await session.executeWorkflow(workflow);
 * 
 * // Stop recording
 * await recording.stop();
 * ```
 */

const axios = require('axios');
const WebSocket = require('ws');
const EventEmitter = require('events');

/**
 * Mouse button types for click operations
 */
const MouseButton = {
  LEFT: 'left',
  RIGHT: 'right', 
  MIDDLE: 'middle'
};

/**
 * Supported desktop environments
 */
const DesktopType = {
  UBUNTU: 'ubuntu',
  UBUNTU_XFCE: 'ubuntu-xfce',
  UBUNTU_KDE: 'ubuntu-kde',
  CENTOS: 'centos',
  FEDORA: 'fedora',
  ARCH: 'arch',
  DEBIAN: 'debian'
};

/**
 * Recording quality presets
 */
const RecordingQuality = {
  LOW: 'low',
  MEDIUM: 'medium',
  HIGH: 'high', 
  STREAMING: 'streaming'
};

/**
 * 2D coordinate point
 */
class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}

/**
 * Information about an active session
 */
class SessionInfo {
  constructor(data) {
    this.sessionId = data.session_id;
    this.userId = data.user_id;
    this.desktopType = data.desktop_type;
    this.status = data.status;
    this.createdAt = data.created_at;
    this.lastActivity = data.last_activity;
    this.recordingActive = data.recording_active;
  }
}

/**
 * Result of workflow execution
 */
class WorkflowResult {
  constructor(data) {
    this.workflowName = data.workflow_name;
    this.success = data.success;
    this.totalSteps = data.total_steps;
    this.successfulSteps = data.successful_steps;
    this.executionTimeMs = data.execution_time_ms;
    this.errors = data.errors || [];
  }
}

/**
 * Individual step in an automation workflow
 */
class WorkflowStep {
  constructor(name, actionType, parameters = {}) {
    this.name = name;
    this.actionType = actionType;
    this.parameters = parameters;
    this.timeoutSeconds = parameters.timeout || 30;
  }
}

/**
 * Automation workflow definition
 */
class Workflow {
  constructor(name, description = '', continueOnError = false) {
    this.name = name;
    this.description = description;
    this.continueOnError = continueOnError;
    this.steps = [];
  }

  /**
   * Add a step to the workflow
   */
  addStep(actionType, parameters = {}, name = null) {
    if (!name) {
      name = `Step ${this.steps.length + 1}: ${actionType}`;
    }
    
    const step = new WorkflowStep(name, actionType, parameters);
    this.steps.push(step);
    return this;
  }

  /**
   * Add cursor movement step
   */
  moveCursor(x, y, name = null) {
    return this.addStep('move_cursor', { x, y }, name);
  }

  /**
   * Add click step
   */
  click(x = null, y = null, button = MouseButton.LEFT, name = null) {
    const params = { button };
    if (x !== null && y !== null) {
      params.x = x;
      params.y = y;
    }
    return this.addStep('click', params, name);
  }

  /**
   * Add text typing step
   */
  typeText(text, name = null) {
    return this.addStep('type', { text }, name);
  }

  /**
   * Add wait/delay step
   */
  wait(seconds, name = null) {
    return this.addStep('wait', { duration: seconds }, name);
  }

  /**
   * Convert workflow to object for API
   */
  toObject() {
    return {
      name: this.name,
      description: this.description,
      continue_on_error: this.continueOnError,
      steps: this.steps.map(step => ({
        name: step.name,
        action_type: step.actionType,
        parameters: step.parameters,
        timeout_seconds: step.timeoutSeconds
      }))
    };
  }
}

/**
 * Recording session control
 */
class Recording extends EventEmitter {
  constructor(session, recordingId) {
    super();
    this.session = session;
    this.recordingId = recordingId;
    this.active = true;
    this.startTime = Date.now();
  }

  /**
   * Stop the recording and return output path
   */
  async stop() {
    if (!this.active) {
      throw new Error('Recording is not active');
    }

    const result = await this.session._apiCall(
      `sessions/${this.session.sessionId}/recording/stop`,
      'POST'
    );

    this.active = false;
    this.emit('stopped', result);
    console.log(`Recording stopped: ${result}`);
    return result;
  }

  /**
   * Get recording duration in seconds
   */
  get duration() {
    return (Date.now() - this.startTime) / 1000;
  }
}

/**
 * Desktop automation session
 */
class Session extends EventEmitter {
  constructor(kvsInstance, sessionId, userId, desktopType) {
    super();
    this.kvs = kvsInstance;
    this.sessionId = sessionId;
    this.userId = userId;
    this.desktopType = desktopType;
    this.currentRecording = null;
    this.websocket = null;
  }

  /**
   * Make API call to KVirtualStage server
   */
  async _apiCall(endpoint, method = 'GET', data = null) {
    return await this.kvs._apiCall(endpoint, method, data);
  }

  /**
   * Move cursor to specified coordinates with natural movement
   */
  async moveCursor(x, y) {
    await this._apiCall(
      `sessions/${this.sessionId}/cursor/move`,
      'POST',
      { target_x: x, target_y: y }
    );
    console.log(`Cursor moved to (${x}, ${y})`);
    this.emit('cursorMoved', { x, y });
  }

  /**
   * Click at current cursor position or specified coordinates
   */
  async click(x = null, y = null, button = MouseButton.LEFT) {
    if (x !== null && y !== null) {
      await this.moveCursor(x, y);
    }

    await this._apiCall(
      `sessions/${this.sessionId}/mouse/click`,
      'POST',
      { button }
    );
    console.log(`Clicked with ${button} button`);
    this.emit('clicked', { x, y, button });
  }

  /**
   * Double-click at current cursor position or specified coordinates
   */
  async doubleClick(x = null, y = null) {
    if (x !== null && y !== null) {
      await this.moveCursor(x, y);
    }

    await this.click();
    await new Promise(resolve => setTimeout(resolve, 100)); // Small delay
    await this.click();
    console.log('Double-clicked');
    this.emit('doubleClicked', { x, y });
  }

  /**
   * Right-click at current cursor position or specified coordinates
   */
  async rightClick(x = null, y = null) {
    await this.click(x, y, MouseButton.RIGHT);
  }

  /**
   * Type text with natural human-like timing
   */
  async typeText(text, wpm = 65) {
    await this._apiCall(
      `sessions/${this.sessionId}/keyboard/type`,
      'POST',
      { text, wpm }
    );
    const displayText = text.length > 50 ? text.substring(0, 50) + '...' : text;
    console.log(`Typed: ${displayText}`);
    this.emit('textTyped', { text });
  }

  /**
   * Press a specific key (e.g., 'Enter', 'Tab', 'Escape')
   */
  async keyPress(key) {
    await this._apiCall(
      `sessions/${this.sessionId}/keyboard/key`,
      'POST',
      { key }
    );
    console.log(`Key pressed: ${key}`);
    this.emit('keyPressed', { key });
  }

  /**
   * Press a combination of keys (e.g., 'Ctrl', 'C')
   */
  async keyCombination(...keys) {
    await this._apiCall(
      `sessions/${this.sessionId}/keyboard/combo`,
      'POST',
      { keys }
    );
    console.log(`Key combination: ${keys.join('+')}`);
    this.emit('keyCombination', { keys });
  }

  /**
   * Scroll in the specified direction
   */
  async scroll(direction, amount = 3) {
    await this._apiCall(
      `sessions/${this.sessionId}/mouse/scroll`,
      'POST',
      { direction, amount }
    );
    console.log(`Scrolled ${direction} by ${amount}`);
    this.emit('scrolled', { direction, amount });
  }

  /**
   * Take a screenshot of the current desktop
   */
  async screenshot(filename = null) {
    const result = await this._apiCall(
      `sessions/${this.sessionId}/screenshot`,
      'POST',
      { filename }
    );
    console.log(`Screenshot taken: ${result}`);
    this.emit('screenshot', { filename: result });
    return result;
  }

  /**
   * Start recording the session
   */
  async startRecording(filename = null, quality = RecordingQuality.MEDIUM) {
    if (this.currentRecording && this.currentRecording.active) {
      throw new Error('Recording is already active');
    }

    if (!filename) {
      filename = `kvs_recording_${Date.now()}.mp4`;
    }

    const result = await this._apiCall(
      `sessions/${this.sessionId}/recording/start`,
      'POST',
      {
        output_filename: filename,
        quality
      }
    );

    const recording = new Recording(this, result.recording_id);
    this.currentRecording = recording;
    console.log(`Recording started: ${filename}`);
    this.emit('recordingStarted', { filename, recordingId: result.recording_id });
    return recording;
  }

  /**
   * Execute an automation workflow
   */
  async executeWorkflow(workflow) {
    const result = await this._apiCall(
      `sessions/${this.sessionId}/workflow`,
      'POST',
      workflow.toObject()
    );

    const workflowResult = new WorkflowResult(result);
    console.log(`Workflow '${workflow.name}' completed: ${workflowResult.success}`);
    this.emit('workflowCompleted', workflowResult);
    return workflowResult;
  }

  /**
   * Connect to WebSocket for live streaming
   */
  connectStream() {
    if (this.websocket) {
      this.websocket.close();
    }

    const wsUrl = `ws://${this.kvs.serverUrl.replace('http://', '').replace('https://', '')}/api/v1/sessions/${this.sessionId}/stream`;
    this.websocket = new WebSocket(wsUrl);

    this.websocket.on('open', () => {
      console.log('WebSocket stream connected');
      this.emit('streamConnected');
    });

    this.websocket.on('message', (data) => {
      try {
        const message = JSON.parse(data);
        this.emit('streamMessage', message);
      } catch (e) {
        this.emit('streamData', data);
      }
    });

    this.websocket.on('close', () => {
      console.log('WebSocket stream disconnected');
      this.emit('streamDisconnected');
    });

    this.websocket.on('error', (error) => {
      console.error('WebSocket error:', error);
      this.emit('streamError', error);
    });

    return this.websocket;
  }

  /**
   * Get detailed session information
   */
  async getInfo() {
    const result = await this._apiCall(`sessions/${this.sessionId}`);
    return new SessionInfo(result);
  }

  /**
   * Close the session and clean up resources
   */
  async close() {
    if (this.currentRecording && this.currentRecording.active) {
      await this.currentRecording.stop();
    }

    if (this.websocket) {
      this.websocket.close();
    }

    await this._apiCall(
      `sessions/${this.sessionId}`,
      'DELETE'
    );
    console.log(`Session closed: ${this.sessionId}`);
    this.emit('closed');
  }
}

/**
 * Main KVirtualStage automation interface
 */
class KVirtualStage extends EventEmitter {
  constructor(serverUrl = 'http://localhost:8080') {
    super();
    this.serverUrl = serverUrl.replace(/\/$/, '');
    this.baseUrl = `${this.serverUrl}/api/v1`;
    this.sessions = new Map();
    this.requestTimeout = 30000; // 30 seconds
  }

  /**
   * Make HTTP API call to KVirtualStage server
   */
  async _apiCall(endpoint, method = 'GET', data = null) {
    const url = `${this.baseUrl}/${endpoint.replace(/^\//, '')}`;
    
    const config = {
      method,
      url,
      timeout: this.requestTimeout,
      headers: {
        'Content-Type': 'application/json'
      }
    };

    if (data) {
      config.data = data;
    }

    try {
      const response = await axios(config);
      
      if (response.data && !response.data.success && response.data.error) {
        throw new Error(`API error: ${response.data.error}`);
      }

      return response.data.data || response.data;
    } catch (error) {
      if (error.response) {
        throw new Error(`API call failed (${error.response.status}): ${error.response.data}`);
      } else if (error.request) {
        throw new Error(`Network error: ${error.message}`);
      } else {
        throw error;
      }
    }
  }

  /**
   * Create a new desktop automation session
   */
  async createSession(options = {}) {
    const {
      userId,
      sessionName = `session_${Date.now()}`,
      desktopType = DesktopType.UBUNTU
    } = options;

    if (!userId) {
      throw new Error('userId is required');
    }

    const result = await this._apiCall(
      'sessions',
      'POST',
      {
        user_id: userId,
        session_name: sessionName,
        desktop_type: desktopType
      }
    );

    const session = new Session(this, result.session_id, userId, desktopType);
    this.sessions.set(session.sessionId, session);

    console.log(`Session created: ${session.sessionId}`);
    this.emit('sessionCreated', session);
    return session;
  }

  /**
   * Get an existing session by ID
   */
  async getSession(sessionId) {
    if (this.sessions.has(sessionId)) {
      return this.sessions.get(sessionId);
    }

    // Try to fetch from server
    try {
      const sessionInfo = await this._apiCall(`sessions/${sessionId}`);
      const session = new Session(this, sessionId, sessionInfo.user_id, sessionInfo.desktop_type);
      this.sessions.set(sessionId, session);
      return session;
    } catch (error) {
      return null;
    }
  }

  /**
   * List all active sessions
   */
  async listSessions() {
    const result = await this._apiCall('sessions');
    return result.map(s => new SessionInfo(s));
  }

  /**
   * Check server health and status
   */
  async healthCheck() {
    return await this._apiCall('health');
  }

  /**
   * Get server performance metrics
   */
  async getMetrics() {
    return await this._apiCall('metrics');
  }

  /**
   * Close all active sessions
   */
  async closeAllSessions() {
    const sessions = Array.from(this.sessions.values());
    await Promise.all(sessions.map(session => session.close()));
    this.sessions.clear();
  }

  /**
   * Set request timeout
   */
  setTimeout(ms) {
    this.requestTimeout = ms;
  }
}

// Export classes and constants
module.exports = {
  KVirtualStage,
  Session,
  Workflow,
  WorkflowStep,
  Recording,
  Point,
  SessionInfo,
  WorkflowResult,
  MouseButton,
  DesktopType,
  RecordingQuality
};

// Also export as ES6 module for modern Node.js
module.exports.default = KVirtualStage;

File: KDesktopVirt/bindings/python/kvirtualstage.py
"""
KVirtualStage Python Bindings

Playwright-equivalent desktop automation platform for AI agents.
Provides high-level Python API for:
- Desktop automation and control
- Session management
- Recording and playback
- Natural human-like interactions
- Cross-platform virtualization

Example Usage:
    >>> import kvirtualstage as kvs
    >>> 
    >>> # Create automation instance
    >>> automation = kvs.KVirtualStage()
    >>> 
    >>> # Create a new desktop session
    >>> session = automation.create_session(
    ...     user_id="demo_user",
    ...     session_name="my_session",
    ...     desktop_type="ubuntu"
    ... )
    >>> 
    >>> # Perform natural automation
    >>> session.move_cursor(400, 300)
    >>> session.click()
    >>> session.type_text("Hello from KVirtualStage!")
    >>> 
    >>> # Start recording
    >>> recording = session.start_recording("demo.mp4")
    >>> 
    >>> # Execute workflow
    >>> workflow = kvs.Workflow("Calculator Demo")
    >>> workflow.add_step("move_cursor", x=100, y=100)
    >>> workflow.add_step("click")
    >>> workflow.add_step("type", text="2 + 2 =")
    >>> session.execute_workflow(workflow)
    >>> 
    >>> # Stop recording
    >>> recording.stop()
"""

import asyncio
import json
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass
from enum import Enum
import time
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class MouseButton(Enum):
    """Mouse button types for click operations."""
    LEFT = "left"
    RIGHT = "right" 
    MIDDLE = "middle"

class DesktopType(Enum):
    """Supported desktop environments."""
    UBUNTU = "ubuntu"
    UBUNTU_XFCE = "ubuntu-xfce"
    UBUNTU_KDE = "ubuntu-kde"
    CENTOS = "centos"
    FEDORA = "fedora"
    ARCH = "arch"
    DEBIAN = "debian"

class RecordingQuality(Enum):
    """Recording quality presets."""
    LOW = "low"
    MEDIUM = "medium" 
    HIGH = "high"
    STREAMING = "streaming"

@dataclass
class Point:
    """2D coordinate point."""
    x: float
    y: float

@dataclass
class SessionInfo:
    """Information about an active session."""
    session_id: str
    user_id: str
    desktop_type: str
    status: str
    created_at: float
    last_activity: float
    recording_active: bool

@dataclass
class WorkflowResult:
    """Result of workflow execution."""
    workflow_name: str
    success: bool
    total_steps: int
    successful_steps: int
    execution_time_ms: int
    errors: List[str]

class WorkflowStep:
    """Individual step in an automation workflow."""
    
    def __init__(self, name: str, action_type: str, **kwargs):
        self.name = name
        self.action_type = action_type
        self.parameters = kwargs
        self.timeout_seconds = kwargs.get('timeout', 30)

class Workflow:
    """Automation workflow definition."""
    
    def __init__(self, name: str, description: str = "", continue_on_error: bool = False):
        self.name = name
        self.description = description
        self.continue_on_error = continue_on_error
        self.steps: List[WorkflowStep] = []
    
    def add_step(self, action_type: str, name: str = None, **kwargs) -> 'Workflow':
        """Add a step to the workflow."""
        if name is None:
            name = f"Step {len(self.steps) + 1}: {action_type}"
        
        step = WorkflowStep(name, action_type, **kwargs)
        self.steps.append(step)
        return self
    
    def move_cursor(self, x: float, y: float, name: str = None) -> 'Workflow':
        """Add cursor movement step."""
        return self.add_step("move_cursor", name=name, x=x, y=y)
    
    def click(self, x: float = None, y: float = None, button: MouseButton = MouseButton.LEFT, name: str = None) -> 'Workflow':
        """Add click step."""
        params = {"button": button.value}
        if x is not None and y is not None:
            params.update({"x": x, "y": y})
        return self.add_step("click", name=name, **params)
    
    def type_text(self, text: str, name: str = None) -> 'Workflow':
        """Add text typing step."""
        return self.add_step("type", name=name, text=text)
    
    def wait(self, seconds: float, name: str = None) -> 'Workflow':
        """Add wait/delay step."""
        return self.add_step("wait", name=name, duration=seconds)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert workflow to dictionary for API."""
        return {
            "name": self.name,
            "description": self.description,
            "continue_on_error": self.continue_on_error,
            "steps": [
                {
                    "name": step.name,
                    "action_type": step.action_type,
                    "parameters": step.parameters,
                    "timeout_seconds": step.timeout_seconds
                }
                for step in self.steps
            ]
        }

class Recording:
    """Recording session control."""
    
    def __init__(self, session: 'Session', recording_id: str):
        self.session = session
        self.recording_id = recording_id
        self.active = True
        self.start_time = time.time()
    
    async def stop(self) -> str:
        """Stop the recording and return output path."""
        if not self.active:
            raise RuntimeError("Recording is not active")
        
        output_path = await self.session._api_call(
            f"sessions/{self.session.session_id}/recording/stop",
            method="POST"
        )
        
        self.active = False
        logger.info(f"Recording stopped: {output_path}")
        return output_path
    
    @property
    def duration(self) -> float:
        """Get recording duration in seconds."""
        return time.time() - self.start_time

class Session:
    """Desktop automation session."""
    
    def __init__(self, kvs_instance: 'KVirtualStage', session_id: str, user_id: str, desktop_type: str):
        self.kvs = kvs_instance
        self.session_id = session_id
        self.user_id = user_id
        self.desktop_type = desktop_type
        self.current_recording: Optional[Recording] = None
    
    async def _api_call(self, endpoint: str, method: str = "GET", data: Dict = None) -> Any:
        """Make API call to KVirtualStage server."""
        return await self.kvs._api_call(endpoint, method, data)
    
    async def move_cursor(self, x: float, y: float) -> None:
        """Move cursor to specified coordinates with natural movement."""
        await self._api_call(
            f"sessions/{self.session_id}/cursor/move",
            method="POST",
            data={"target_x": x, "target_y": y}
        )
        logger.info(f"Cursor moved to ({x}, {y})")
    
    async def click(self, x: float = None, y: float = None, button: MouseButton = MouseButton.LEFT) -> None:
        """Click at current cursor position or specified coordinates."""
        if x is not None and y is not None:
            await self.move_cursor(x, y)
        
        await self._api_call(
            f"sessions/{self.session_id}/mouse/click",
            method="POST",
            data={"button": button.value}
        )
        logger.info(f"Clicked with {button.value} button")
    
    async def double_click(self, x: float = None, y: float = None) -> None:
        """Double-click at current cursor position or specified coordinates."""
        if x is not None and y is not None:
            await self.move_cursor(x, y)
        
        await self.click()
        await asyncio.sleep(0.1)  # Small delay between clicks
        await self.click()
        logger.info("Double-clicked")
    
    async def right_click(self, x: float = None, y: float = None) -> None:
        """Right-click at current cursor position or specified coordinates."""
        await self.click(x, y, MouseButton.RIGHT)
    
    async def type_text(self, text: str, wpm: float = 65.0) -> None:
        """Type text with natural human-like timing."""
        await self._api_call(
            f"sessions/{self.session_id}/keyboard/type",
            method="POST",
            data={"text": text, "wpm": wpm}
        )
        logger.info(f"Typed: {text[:50]}{'...' if len(text) > 50 else ''}")
    
    async def key_press(self, key: str) -> None:
        """Press a specific key (e.g., 'Enter', 'Tab', 'Escape')."""
        await self._api_call(
            f"sessions/{self.session_id}/keyboard/key",
            method="POST", 
            data={"key": key}
        )
        logger.info(f"Key pressed: {key}")
    
    async def key_combination(self, *keys: str) -> None:
        """Press a combination of keys (e.g., 'Ctrl', 'C')."""
        await self._api_call(
            f"sessions/{self.session_id}/keyboard/combo",
            method="POST",
            data={"keys": list(keys)}
        )
        logger.info(f"Key combination: {'+'.join(keys)}")
    
    async def scroll(self, direction: str, amount: int = 3) -> None:
        """Scroll in the specified direction."""
        await self._api_call(
            f"sessions/{self.session_id}/mouse/scroll",
            method="POST",
            data={"direction": direction, "amount": amount}
        )
        logger.info(f"Scrolled {direction} by {amount}")
    
    async def screenshot(self, filename: str = None) -> str:
        """Take a screenshot of the current desktop."""
        result = await self._api_call(
            f"sessions/{self.session_id}/screenshot",
            method="POST",
            data={"filename": filename}
        )
        logger.info(f"Screenshot taken: {result}")
        return result
    
    async def start_recording(self, filename: str = None, quality: RecordingQuality = RecordingQuality.MEDIUM) -> Recording:
        """Start recording the session."""
        if self.current_recording and self.current_recording.active:
            raise RuntimeError("Recording is already active")
        
        if filename is None:
            filename = f"kvs_recording_{int(time.time())}.mp4"
        
        result = await self._api_call(
            f"sessions/{self.session_id}/recording/start",
            method="POST",
            data={
                "output_filename": filename,
                "quality": quality.value
            }
        )
        
        recording = Recording(self, result["recording_id"])
        self.current_recording = recording
        logger.info(f"Recording started: {filename}")
        return recording
    
    async def execute_workflow(self, workflow: Workflow) -> WorkflowResult:
        """Execute an automation workflow."""
        result = await self._api_call(
            f"sessions/{self.session_id}/workflow",
            method="POST",
            data=workflow.to_dict()
        )
        
        workflow_result = WorkflowResult(
            workflow_name=result["workflow_name"],
            success=result["success"],
            total_steps=result["total_steps"],
            successful_steps=result["successful_steps"],
            execution_time_ms=result["execution_time_ms"],
            errors=result["errors"]
        )
        
        logger.info(f"Workflow '{workflow.name}' completed: {workflow_result.success}")
        return workflow_result
    
    async def get_info(self) -> SessionInfo:
        """Get detailed session information."""
        result = await self._api_call(f"sessions/{self.session_id}")
        
        return SessionInfo(
            session_id=result["session_id"],
            user_id=result["user_id"], 
            desktop_type=result["desktop_type"],
            status=result["status"],
            created_at=result.get("created_at", 0),
            last_activity=result.get("last_activity", 0),
            recording_active=result.get("recording_active", False)
        )
    
    async def close(self) -> None:
        """Close the session and clean up resources."""
        if self.current_recording and self.current_recording.active:
            await self.current_recording.stop()
        
        await self._api_call(
            f"sessions/{self.session_id}",
            method="DELETE"
        )
        logger.info(f"Session closed: {self.session_id}")

class KVirtualStage:
    """Main KVirtualStage automation interface."""
    
    def __init__(self, server_url: str = "http://localhost:8080"):
        self.server_url = server_url.rstrip('/')
        self.base_url = f"{self.server_url}/api/v1"
        self.sessions: Dict[str, Session] = {}
    
    async def _api_call(self, endpoint: str, method: str = "GET", data: Dict = None) -> Any:
        """Make HTTP API call to KVirtualStage server."""
        import aiohttp
        
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        
        async with aiohttp.ClientSession() as session:
            kwargs = {
                "headers": {"Content-Type": "application/json"}
            }
            
            if data:
                kwargs["json"] = data
            
            async with session.request(method, url, **kwargs) as response:
                if response.status >= 400:
                    text = await response.text()
                    raise RuntimeError(f"API call failed ({response.status}): {text}")
                
                result = await response.json()
                
                if not result.get("success", True):
                    error_msg = result.get("error", "Unknown error")
                    raise RuntimeError(f"API error: {error_msg}")
                
                return result.get("data", result)
    
    async def create_session(self, user_id: str, session_name: str = None, 
                           desktop_type: DesktopType = DesktopType.UBUNTU) -> Session:
        """Create a new desktop automation session."""
        if session_name is None:
            session_name = f"session_{int(time.time())}"
        
        result = await self._api_call(
            "sessions",
            method="POST",
            data={
                "user_id": user_id,
                "session_name": session_name,
                "desktop_type": desktop_type.value
            }
        )
        
        session = Session(self, result["session_id"], user_id, desktop_type.value)
        self.sessions[session.session_id] = session
        
        logger.info(f"Session created: {session.session_id}")
        return session
    
    async def get_session(self, session_id: str) -> Optional[Session]:
        """Get an existing session by ID."""
        if session_id in self.sessions:
            return self.sessions[session_id]
        
        # Try to fetch from server
        try:
            session_info = await self._api_call(f"sessions/{session_id}")
            session = Session(self, session_id, session_info["user_id"], session_info["desktop_type"])
            self.sessions[session_id] = session
            return session
        except RuntimeError:
            return None
    
    async def list_sessions(self) -> List[SessionInfo]:
        """List all active sessions."""
        result = await self._api_call("sessions")
        
        return [
            SessionInfo(
                session_id=s["session_id"],
                user_id=s["user_id"],
                desktop_type=s["desktop_type"],
                status=s["status"],
                created_at=s.get("created_at", 0),
                last_activity=s.get("last_activity", 0),
                recording_active=s.get("recording_active", False)
            )
            for s in result
        ]
    
    async def health_check(self) -> Dict[str, Any]:
        """Check server health and status."""
        return await self._api_call("health")
    
    async def get_metrics(self) -> Dict[str, Any]:
        """Get server performance metrics."""
        return await self._api_call("metrics")
    
    async def close_all_sessions(self) -> None:
        """Close all active sessions."""
        for session in list(self.sessions.values()):
            await session.close()
        self.sessions.clear()

# Convenience functions for synchronous usage
def run_async(coro):
    """Run an async function synchronously."""
    try:
        loop = asyncio.get_event_loop()
    except RuntimeError:
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
    
    return loop.run_until_complete(coro)

class SyncSession:
    """Synchronous wrapper for Session class."""
    
    def __init__(self, async_session: Session):
        self._async_session = async_session
    
    def move_cursor(self, x: float, y: float) -> None:
        return run_async(self._async_session.move_cursor(x, y))
    
    def click(self, x: float = None, y: float = None, button: MouseButton = MouseButton.LEFT) -> None:
        return run_async(self._async_session.click(x, y, button))
    
    def type_text(self, text: str, wpm: float = 65.0) -> None:
        return run_async(self._async_session.type_text(text, wpm))
    
    def screenshot(self, filename: str = None) -> str:
        return run_async(self._async_session.screenshot(filename))
    
    def execute_workflow(self, workflow: Workflow) -> WorkflowResult:
        return run_async(self._async_session.execute_workflow(workflow))
    
    def close(self) -> None:
        return run_async(self._async_session.close())

class SyncKVirtualStage:
    """Synchronous wrapper for KVirtualStage class."""
    
    def __init__(self, server_url: str = "http://localhost:8080"):
        self._async_kvs = KVirtualStage(server_url)
    
    def create_session(self, user_id: str, session_name: str = None,
                      desktop_type: DesktopType = DesktopType.UBUNTU) -> SyncSession:
        async_session = run_async(self._async_kvs.create_session(user_id, session_name, desktop_type))
        return SyncSession(async_session)
    
    def list_sessions(self) -> List[SessionInfo]:
        return run_async(self._async_kvs.list_sessions())
    
    def health_check(self) -> Dict[str, Any]:
        return run_async(self._async_kvs.health_check())

# Export public API
__all__ = [
    'KVirtualStage', 'SyncKVirtualStage',
    'Session', 'SyncSession', 
    'Workflow', 'WorkflowStep', 'Recording',
    'MouseButton', 'DesktopType', 'RecordingQuality',
    'Point', 'SessionInfo', 'WorkflowResult',
    'run_async'
]

# Version info
__version__ = "0.1.0"
__author__ = "KVirtualStage Team"
__description__ = "Playwright-equivalent desktop automation platform for AI agents"